src/view/ChatView.tsx

```diff
@@ -95,6 +100,20 @@ export default function ChatView() {
 
       try {
         const text = await read_file_as_text(file);
+
+        const ds = DataSources[DataSources.length - 1] as CustomDataSource;
+        const customChars = await deserialize_custom_chars(text, ds)
+        const chars = new Map(ctx.characters)
+        customChars.map((char) => {
+          const ch = chars.get(char.id);
+          if (ch != null) {
+            chars.delete(char.id)
+          }
+          chars.set(char.id, char);
+        })
+        ctx.characters = chars;
```

Comment:
Don't modify `ctx`. Change param on L117 instead.

Also consider skipping the update if no custom chars are imported.